Add documentation for Github OIDC federation with Keycloak (#181)
securesign/sigstore-ocp: configure-oidc.md

# OIDC configuration

> While RHTAS uses the RHSSO operator (based on Keycloak) for OIDC by default, other OIDC options are available. Integrating with these options is described below. These steps should be followed after [RHTAS installation](https://access.redhat.com/documentation/en-us/red_hat_trusted_artifact_signer/2024-q1/html/deployment_guide/index), as they will assume that other components, including the Keycloak client, are operational, and that the environment variables for Cosign initialization are [populated](https://github.com/securesign/sigstore-ocp/blob/main/sign-verify.md#signing-a-container-from-the-local-system).

Reviewer: ```suggestion
While RHTAS uses the RHSSO operator (based on Keycloak) for OIDC by default, other OIDC options are available. Integrating with these options is described below. These steps should be followed after [RHTAS installation](https://access.redhat.com/documentation/en-us/red_hat_trusted_artifact_signer/2024-q1/html/deployment_guide/index), as they will assume that other components, including the Keycloak client for RHTAS, is present, and that the environment variables for Cosign initialization are [populated](https://github.com/securesign/sigstore-ocp/blob/main/sign-verify.md#signing-a-container-from-the-local-system). You may need to recreate the Keycloak client for RHTAS once the `sigstore` realm has been fully configured as described within this guide.
```

> 

Author: @sabre1041 I added this to point them to TAS installation steps.

### Update Fulcio deployment configuration

> Once [RHTAS is installed](https://github.com/securesign/sigstore-ocp/blob/main/tas-easy-install.sh), issue

Author: That's fair. What if we point them to the deployment guide instead? https://access.redhat.com/documentation/en-us/red_hat_trusted_artifact_signer/2024-q1/html/deployment_guide/installing-red-hat-trusted-artifact-signer-on-openshift_deploy

> Once [RHTAS is installed](https://github.com/securesign/sigstore-ocp/blob/main/tas-easy-install.sh), issue

Author: I'm actually going to remove this change in this PR as it's GCP related, but the same applies for the Github installation step.

```
cosign sign -y --fulcio-url=$FULCIO_URL --rekor-url=$REKOR_URL --oidc-issuer=$OIDC_ISSUER_URL $IMAGE --oidc-client-secret-file=<my secret filename> --oidc-client-id=<my client ID>
```

## Github (federated via Keycloak)

### Obtain Client ID and Secret from Github OAuth App

To integrate RHTAS with Github as an identity provider, you must register your Github OAuth App in your Keycloak realm. 

Locate your Github Oauth App, or create a new Github Oauth App following the directions [here](https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/creating-an-oauth-app). When creating the OAuth App, it is fine to enter a placeholder value for the Authorization Callback URL field, as this field will need to be updated to hold a value from Keycloak after more setup steps are complete. The Homepage URL field can be any value.

Locate the client ID and client secret values from the Oauth App page, as they will be required to create the custom resource file.

### Log in to Keycloak console

> You will need to log in to the Keycloak console to complete some steps of this process. Once [RHTAS is installed](https://github.com/securesign/sigstore-ocp/blob/main/tas-easy-install.sh), you can do this by finding the route and admin password.

Author: Going to still update this with a link TBD

```shell
$ oc get routes -n keycloak-system | awk '/keycloak/ {print $2}' | head -1
keycloak-keycloak-system.apps.rosa.umebj-fwkt3-8j8.y6tr.p3.openshiftapps.com
```

> Copy this route, adding `https://` in front, into your browser.

Author: Do you mean the output of the command or the note about https? I've reworked it so please take a look (will push the new version shortly). I do think it's important to make a note about https in case they haven't done this before, as otherwise it will not work.

> Copy this route, adding `https://` in front, into your browser.

Author: Oh I see, your command adds that already. Neat! :tada:

You will need to log in. The username will be `admin`. You can retrieve the password with:
```shell

> $ oc get secret/credential-keycloak -n keycloak-system -o yaml | awk '/ADMIN_PASSWORD:/ {print $2}' | base64 -d; printf "\n"

Author: I like it. Though I do prefer the newline, I can also see the argument against it for a cleaner command.

```

### Add Github Identity Provider to Keycloak Realm

Select one of the options below to set up identity federation to the Github Identity Provider. 

#### Option 1: Create a KeycloakRealm in OpenShift on the CLI

> Create a Custom Resource file from which to create the KeycloakRealm that includes the Github identity provider:

Author: Well, they would need to delete it, as far as I'm aware (described below), or create it with a different name.

> Create a Custom Resource file from which to create the KeycloakRealm that includes the Github identity provider:

Author: I'm going through a flow where they make one with a new name right now, to test that, but I think it is actually easier if they delete the existing one. I get this error
```
status:
  loginURL: ""
  message: 'failed to create realm: (409) 409 Conflict'
  phase: failing
  ready: false
```
on the new Keycloakrealm when I haven't deleted the pre-existing `sigstore` one - I think because the IDP already exists in a different Keycloakrealm, but I am not sure.

```shell
$ cat realm-cr.yaml
apiVersion: keycloak.org/v1alpha1
kind: KeycloakRealm
metadata:
  labels:
    app: sso
  name: sigstore
spec:
  instanceSelector:
    matchLabels:
      app: sso
  realm:
    displayName: Sigstore
    enabled: true
    id: sigstore
    realm: sigstore
    sslRequired: none
    identityProviders:
      - alias: github
        displayName: Github
        enabled: true
        firstBrokerLoginFlowAlias: first broker login
        internalId: github
        providerId: github
        trustEmail: true
        config:

>             clientId: <my client id> # add your client ID

Author: Thanks, the spacing was hard to get right. Let me know if there's still something wrong with it in the new version

            clientSecret: <my client secret> # add your client secret
```

Make sure that you have no existing KeycloakRealm with the `metadata.name` specified in your `realm-cr.yaml` file. IMPORTANT: make sure you are okay with deleting your current `sigstore` KeycloakRealm.
```shell
$ oc delete keycloakrealm sigstore -n keycloak-system

> keycloakrealm.keycloak.org "sigstore" deleted

Author: Not sure, see above

> keycloakrealm.keycloak.org "sigstore" deleted

Author: Oh I see, you mean the output

```

Create the new KeycloakRealm that includes the Github identity provider from the custom resource file:
```shell
$ oc create -f realm-cr.yaml -n keycloak-system
keycloakrealm.keycloak.org/sigstore created
```

#### Option 2: 

> You may also add the Github Identity provider to an existing KeycloakRealm through the Keycloak console by selecting `Identity Providers` in the lefthand menu bar and selecting `Add provider -> Github` from the drop down menu. See above for instructions on how to access the Keycloak console.

Author: I've added a few more steps here.

### Copy the Redirect URI to your Github OAuth App

From the Keycloak console, navigate to `Identity Providers` in the lefthand menu bar, and select the provider you have just created: `github`. It should look like this:

<p align="center">
  <img src="/images/github_idp.png" />
</p>

From this page, copy the Redirect URI:

<p align="center">
  <img src="/images/redirect_uri.png" />
</p>

Paste the redirect URI value into the OAuth app Authorization Callback URL field, then click `Update Application`:
<p align="center">
  <img src="/images/github_callback.png" />
</p>

> ### Add IdentityProviderMapper

Reviewer: Can we configure this via the `KeycloakRealm` CR to automate the process?

Author: That's what I was trying to figure out. As far as I could tell, there was no way to do it with the RHSSO Operator. Adding an `IdentityProviderMapper` or `IdentityProviderMappers` key was either ignored or made the KeycloakRealm fail to come up. If you know of a way to do this, I would love to add it.

> From the Keycloak console, navigate to `Identity Providers` in the lefthand menu bar, and select the provider you have just created: `github`. Select the tab `Mappers` and on the right-hand side, click the `create` button.

Reviewer: ```suggestion
From the Keycloak console, navigate to `Identity Providers` in the lefthand menu bar, and select the provider you have just created: `github`. Select the tab `Mappers` and on the right-hand side and click the `create` button.
```

> Set the field values as shown below. The 'name' can be anything, as this will be overwritten.

Reviewer: Recommend providing the values here as well

<p align="center">
  <img src="/images/mapper2.png" />
</p>

Make sure to hit `save`.

### Sign the artifact

> Sign the artifact using the same commands listed in [`sign-verify.md`](https://github.com/securesign/sigstore-ocp/blob/release-1.0.beta/sign-verify.md):

Reviewer: Recommend using relative pathing which will always align to the branch that the file is currently on

```shell
$ cosign initialize --mirror=$TUF_URL --root=$TUF_URL/root.json
$ cosign sign -y --fulcio-url=$FULCIO_URL --rekor-url=$REKOR_URL --oidc-issuer=$OIDC_ISSUER_URL $IMAGE
```

> On the login page the pops up in the browser, make sure to select the Github login option at the bottom.

Reviewer: ```suggestion
On the login page the pops up in the browser, make sure to select the GitHub login option at the bottom.
```

  <img src="/images/mapper2.png" />
</p>

Make sure to hit `save`.

### Copy the Redirect URI to your GitHub OAuth App

From the Keycloak console, navigate to `Identity Providers` in the lefthand menu bar, and select the provider you have just created: `github`. From this page, copy the Redirect URI:

<p align="center">
  <img src="/images/redirect_uri.png" />
</p>

Paste the redirect URI value into the OAuth app Authorization Callback URL field. Then paste it into the Homepage URL field, removing `broker/github/endpoint` from the end of the URL.

The values should look like this, where `CLUSTERVALUE` is the unique value from the URL of your OpenShift cluster:

|Field                       | Value|
|----------------------------|------|
| Homepage URL               | https://keycloak-keycloak-system.apps.rosa.CLUSTERVALUE.p3.openshiftapps.com/auth/realms/sigstore/ |
| Authorization Callback URL | https://keycloak-keycloak-system.apps.rosa.CLUSTERVALUE.p3.openshiftapps.com/auth/realms/sigstore/broker/github/endpoint |

Add these to the appropriate fields marked below:

<p align="center">
  <img src="/images/add_urls.png" />
</p>

Make sure to click `Update Application`.

> ### Sign the artifact

Reviewer: Before signing the artifact, the `sigstore` client must be available within the `sigstore` realm. If you are creating the `sigstore` realm from scratch, it will not have the `sigstore` client unless you previously installed RHTAS which included both the realm and client and then created the `sigstore` realm via the CLI. it would then get recreated by the operator.

Regardless, there does need to be some wording that describes steps that the user needs to take regardless of the approach

Author: This is a good point, I'm putting a note in about the assumptions re: installation and will ping to see what you think.

Sign the artifact using the same commands listed in [`sign-verify.md`](/sign-verify.md):

```shell

> $ cosign initialize --mirror=$TUF_URL --root=$TUF_URL/root.json

Author: It resolves to this https://github.com/securesign/sigstore-ocp/blob/main/tas-easy-install.sh#L67 as long as they're running the `./tas-easy-install.sh` script. I agree we should either make that assumption explicit or give them other steps to take

> $ cosign initialize --mirror=$TUF_URL --root=$TUF_URL/root.json

Reviewer: The end user still needs to run `source ./tas-env-variables.sh` as described upon completion of the `./tas-easy-install.sh` script

> $ cosign initialize --mirror=$TUF_URL --root=$TUF_URL/root.json

Author: Setting up the env vars is described just above that by linking out to the `sign-verify.md` doc, but I've just added another line to say that they can also do it by using the post-install script.

</p>

Paste the redirect URI value into the OAuth app Authorization Callback URL field. Then paste it into the Homepage URL field, removing `broker/github/endpoint` from the end of the URL.

The values should look like this, where `CLUSTERVALUE` is the unique value from the URL of your OpenShift cluster:

|Field                       | Value|
|----------------------------|------|
| Homepage URL               | https://keycloak-keycloak-system.apps.rosa.CLUSTERVALUE.p3.openshiftapps.com/auth/realms/sigstore/ |
| Authorization Callback URL | https://keycloak-keycloak-system.apps.rosa.CLUSTERVALUE.p3.openshiftapps.com/auth/realms/sigstore/broker/github/endpoint |

Add these to the appropriate fields marked below:

<p align="center">
  <img src="/images/add_urls.png" />
</p>

Make sure to click `Update Application`.

### Sign the artifact

Follow the commands listed in [`sign-verify.md`](/sign-verify.md) to ensure that your TAS environment variables are populated, then initialize Cosign and sign the artifact.
**Troubleshooting:** If you have initialized Cosign previously, you may need to first delete `~/.sigstore` before running the commands below.

> 

Author: @sabre1041 I updated this as well to specify how to populate the env vars.
